pkg/promutil/implement.go

```diff
@@ -0,0 +1,119 @@
+// Registerer. Panic if it can't register successfully. Thread-safe.
+func (f *wrappingFactory) NewCounter(opts prometheus.CounterOpts) prometheus.Counter {
+	c := prometheus.NewCounter(*wrapCounterOpts(f.prefix, f.constLabels, &opts))
+	f.r.MustRegister(f.id, c)
+	return c
+}
+
+// NewCounterVec works like the function of the same name in the
+// prometheus, package but it automatically registers the CounterVec with
+// the Factory's Registerer. Panic if it can't register successfully.Thread-safe.
+func (f *wrappingFactory) NewCounterVec(opts prometheus.CounterOpts, labelNames []string) *prometheus.CounterVec {
+	c := prometheus.NewCounterVec(*wrapCounterOpts(f.prefix, f.constLabels, &opts), labelNames)
+	f.r.MustRegister(f.id, c)
+	return c
+}
+
+// NewGauge works like the function of the same name in the prometheus
+// package, but it automatically registers the Gauge with the Factory's
+// Registerer. Panic if it can't register successfully.Thread-safe.
+func (f *wrappingFactory) NewGauge(opts prometheus.GaugeOpts) prometheus.Gauge {
+	c := prometheus.NewGauge(*wrapGaugeOpts(f.prefix, f.constLabels, &opts))
+	f.r.MustRegister(f.id, c)
+	return c
+}
+
+// NewGaugeVec works like the function of the same name in the prometheus
+// package but it automatically registers the GaugeVec with the Factory's
+// Registerer. Panic if it can't register successfully.Thread-safe.
+func (f *wrappingFactory) NewGaugeVec(opts prometheus.GaugeOpts, labelNames []string) *prometheus.GaugeVec {
+	c := prometheus.NewGaugeVec(*wrapGaugeOpts(f.prefix, f.constLabels, &opts), labelNames)
+	f.r.MustRegister(f.id, c)
+	return c
+}
+
+// NewHistogram works like the function of the same name in the prometheus
+// package but it automatically registers the Histogram with the Factory's
+// Registerer. Panic if it can't register successfully.Thread-safe.
+func (f *wrappingFactory) NewHistogram(opts prometheus.HistogramOpts) prometheus.Histogram {
+	c := prometheus.NewHistogram(*wrapHistogramOpts(f.prefix, f.constLabels, &opts))
+	f.r.MustRegister(f.id, c)
+	return c
+}
+
+// NewHistogramVec works like the function of the same name in the
+// prometheus package but it automatically registers the HistogramVec
+// with the Factory's Registerer. Panic if it can't register successfully.Thread-safe.
+func (f *wrappingFactory) NewHistogramVec(opts prometheus.HistogramOpts, labelNames []string) *prometheus.HistogramVec {
+	c := prometheus.NewHistogramVec(*wrapHistogramOpts(f.prefix, f.constLabels, &opts), labelNames)
+	f.r.MustRegister(f.id, c)
+	return c
+}
+
+func wrapCounterOpts(prefix string, constLabels prometheus.Labels, opts *prometheus.CounterOpts) *prometheus.CounterOpts {
+	if prefix != "" {
+		opts.Namespace = prefix + "_" + opts.Namespace
+	}
+	cls := opts.ConstLabels
+	for name, value := range constLabels {
+		if _, exists := cls[name]; exists {
+			log.L().Panic("duplicate label name", zap.String("label", name))
+		}
+		cls[name] = value
+	}
```

Comment:
We can abstract this logic into a function and reuse it.

Reply:
solve it in last commit